runtime/src/comm/ugni/comm-ugni.c

```diff
@@ -2807,67 +2812,60 @@ void chpl_comm_impl_regMemPostAlloc(void* p, size_t size)
 
-  mem_regions_lock();
-
   //
-  // Finish filling the table entry and register the memory.
+  // Fill in the length, which we set to 1 earlier to avoid lookups
+  // for transfers from matching this entry.
   //
   mr->len = (uint64_t) size;
 
+  //
+  // Register the memory.
+  //
   register_mem_region(mr);
 
   //
-  // Update the copies of our memory regions on all nodes.  If the new
-  // entry doesn't extend our table then we just send the entry.  If it
-  // does, then we need to send both the new count and the new entry.
-  // And so we can do it with just one PUT, we also send everything in
-  // between.
+  // Update the copies of our memory regions on all nodes.  If this
+  // entry is within the already-known range of our table entries then
+  // send only this one.  In this case we don't even need to hold the
+  // critical section lock, because nobody else can be communicating
+  // this entry.  Otherwise, send all the entries beyond the already
+  // known range and then send the new count.  This will probably
+  // include some new unregistered entries for which we have an address
+  // but no length or MDH.  But the presence of those won't create
+  // confusion because the remote nodes won't by trying to address
+  // within them yet anyway, and later when we do register them we may
+  // be able to send just the entries and not the count again.  In this
+  // case we have to hold the lock until the broadcast is complete.
   //
-  {
-    size_t off;
-    size_t size;
-    void* src_v[MAX_CHAINED_PUT_LEN];
-    int32_t node_v[MAX_CHAINED_PUT_LEN];
-    void* tgt_v[MAX_CHAINED_PUT_LEN];
-    size_t size_v[MAX_CHAINED_PUT_LEN];
-    int ci;
-
-    if (mem_regions.mreg_cnt == mem_regions_map[chpl_nodeID].mreg_cnt) {
-      off = (char*) mr - (char*) &mem_regions;
-      size = sizeof(*mr);
-    } else {
-      assert(mem_regions.mreg_cnt
-             > mem_regions_map[chpl_nodeID].mreg_cnt);
-      off = 0;
-      size = (char*) &mem_regions.mregs[mr_i + 1] - (char*) &mem_regions;
-    }
-
-    ci = 0;
-    for (int ni = 0; ni < (int) chpl_numNodes; ni++) {
-      if (ci == MAX_CHAINED_PUT_LEN) {
-        do_remote_put_V(ci, src_v, node_v, tgt_v, size_v, true);
-        ci = 0;
-      }
+  // NB: Is there a "tearing" issue here?  In particular, could a search
+  //     on a remote node see an address that was only partly delivered
+  //     and thereby appeared to specify some other range?
+  //
+  mem_regions_lock();
 
-      if (ni == chpl_nodeID) {
-        memcpy((char*) &mem_regions_map_addr_map[ni][chpl_nodeID] + off,
-               (char*) &mem_regions + off,
-               size);
-      } else {
-        src_v[ci] = (char*) &mem_regions + off;
-        node_v[ci] = ni;
-        tgt_v[ci] = (char*) &mem_regions_map_addr_map[ni][chpl_nodeID] + off;
-        size_v[ci] = size;
-        ci++;
-      }
-    }
+  if (mr_i < mem_regions_map[chpl_nodeID].mreg_cnt) {
+    mem_regions_unlock();
```

Comment:
Found myself wondering post-review -- even if we don't need the mutex lock, don't we still need to prevent yields here?

Reply:
We don't need it as long as we're just dealing with array memory, but yes, in the dynamic stack extension case we'll need it.  Good catch!

Reply:
Depending on the perf results, we probably want to split out the lock and yield, so that we can release the mutex, but continue to not yield